src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/FrameRequestStream.cs

```diff
@@ -105,11 +107,26 @@ private Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationTo
         public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
             var task = ValidateState(cancellationToken);
-            if (task == null)
+            if (task != null)
             {
-                return _body.ReadAsync(new ArraySegment<byte>(buffer, offset, count), cancellationToken);
+                return task;
+            }
+
+            return ReadAsyncInternal(buffer, offset, count, cancellationToken);
+        }
+
+        private async Task<int> ReadAsyncInternal(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            try
+            {
+                return await _body.ReadAsync(new ArraySegment<byte>(buffer, offset, count), cancellationToken);
+            }
+            catch (ConnectionAbortedException)
+            {
+                ExceptionDispatchInfo.Capture(new TaskCanceledException("The request was aborted")).Throw();
```

Comment:
`ExceptionDispatchInfo.Capture` makes sense for preserving the original stack trace, bus since there isn't one why not just `throw new TaskCanceledException("The request was aborted")`? This way you can remove the deceptive `return 0`.